## Prediction
Between the two-prediction models (Logistic Regression and Random Forest Classifier) given to train and test the sample datasets, I anticipate Logistic Regression model to perform with better results given the fact that logistic regression model is well-designed to predict discrete set of classes which is in this case what the we are trying to accomplish through given data to analyze to predict the loan risk rate whether it’s high or low. 
Also, considering that the data set is using unscaled data, logistic regression model will be more sensitive to each variable with higher value and will provide biased results. 

However, with the scaled data Logistic Regression model may provide better results given the fact that this exercise is looking to create a prediction model with discrete classes, which in this case is looking to determine whether the loan is at high or low risk. In addition, most of the features can be classified based on it’s value such as, high vs low loan amount, high income vs low income, and others. 


In [1]:
import numpy as np
import pandas as pd
from pathlib import Path
from sklearn.preprocessing import StandardScaler, LabelEncoder
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import mean_squared_error, r2_score

In [2]:
train_df = pd.read_csv(Path('Resources/2019loans.csv'),index_col=False)
test_df = pd.read_csv(Path('Resources/2020Q1loans.csv'),index_col=False)

In [3]:
train_df.head()

loan_amnt  int_rate  installment home_ownership  annual_inc  \
0     7000.0    0.1894       256.38       MORTGAGE     75000.0   
1    40000.0    0.1614       975.71       MORTGAGE    102000.0   
2    11000.0    0.2055       294.81           RENT     45000.0   
3     4000.0    0.1612       140.87       MORTGAGE     38000.0   
4    14000.0    0.1797       505.93       MORTGAGE     43000.0   

  verification_status pymnt_plan    dti  delinq_2yrs  inq_last_6mths  ...  \
0        Not Verified          n  28.62          0.0             2.0  ...   
1     Source Verified          n  11.72          2.0             0.0  ...   
2            Verified          n  37.25          1.0             3.0  ...   
3        Not Verified          n  42.89          1.0             0.0  ...   
4     Source Verified          n  22.16          1.0             0.0  ...   

   percent_bc_gt_75  pub_rec_bankruptcies  tax_liens  tot_hi_cred_lim  \
0              87.5                   0.0        0.0         352260.0   
1               0.0                   0.0        0.0         294664.0   
2               7.7                   0.0        0.0          92228.0   
3             100.0                   0.0        0.0         284273.0   
4              25.0                   0.0        0.0         120280.0   

  total_bal_ex_mort  total_bc_limit  total_il_high_credit_limit  \
0           62666.0         35000.0                     10000.0   
1          109911.0          9000.0                     71044.0   
2           36007.0         33000.0                     46328.0   
3           52236.0         13500.0                     52017.0   
4           88147.0         33300.0                     78680.0   

   hardship_flag  debt_settlement_flag    target  
0              N                     N  low_risk  
1              N                     N  low_risk  
2              N                     N  low_risk  
3              N                     N  low_risk  
4              N                     N  low_risk  

[5 rows x 84 columns]

In [4]:
X_train_df = train_df.drop('target', axis=1)
X_test_df = test_df.drop('target', axis=1)
X_train_df

loan_amnt  int_rate  installment home_ownership  annual_inc  \
0         7000.0    0.1894       256.38       MORTGAGE     75000.0   
1        40000.0    0.1614       975.71       MORTGAGE    102000.0   
2        11000.0    0.2055       294.81           RENT     45000.0   
3         4000.0    0.1612       140.87       MORTGAGE     38000.0   
4        14000.0    0.1797       505.93       MORTGAGE     43000.0   
...          ...       ...          ...            ...         ...   
12175    19975.0    0.2565       801.09           RENT     28000.0   
12176    15000.0    0.1774       540.34           RENT     50000.0   
12177     3600.0    0.1862       131.28           RENT     60000.0   
12178    15000.0    0.0881       475.68       MORTGAGE     62000.0   
12179    15000.0    0.1774       540.34           RENT     54080.0   

      verification_status pymnt_plan    dti  delinq_2yrs  inq_last_6mths  ...  \
0            Not Verified          n  28.62          0.0             2.0  ...   
1         Source Verified          n  11.72          2.0             0.0  ...   
2                Verified          n  37.25          1.0             3.0  ...   
3            Not Verified          n  42.89          1.0             0.0  ...   
4         Source Verified          n  22.16          1.0             0.0  ...   
...                   ...        ...    ...          ...             ...  ...   
12175        Not Verified          n  28.42          0.0             0.0  ...   
12176            Verified          n  23.43          4.0             0.0  ...   
12177        Not Verified          n  28.80          0.0             1.0  ...   
12178     Source Verified          n  11.44          0.0             0.0  ...   
12179        Not Verified          n  21.66          1.0             1.0  ...   

       pct_tl_nvr_dlq  percent_bc_gt_75  pub_rec_bankruptcies  tax_liens  \
0               100.0              87.5                   0.0        0.0   
1                85.7               0.0                   0.0        0.0   
2                94.3               7.7                   0.0        0.0   
3                92.9             100.0                   0.0        0.0   
4                84.1              25.0                   0.0        0.0   
...               ...               ...                   ...        ...   
12175           100.0              16.7                   0.0        0.0   
12176            90.5              11.1                   0.0        0.0   
12177           100.0               0.0                   0.0        0.0   
12178           100.0               0.0                   0.0        0.0   
12179            90.0             100.0                   0.0        0.0   

      tot_hi_cred_lim  total_bal_ex_mort  total_bc_limit  \
0            352260.0            62666.0         35000.0   
1            294664.0           109911.0          9000.0   
2             92228.0            36007.0         33000.0   
3            284273.0            52236.0         13500.0   
4            120280.0            88147.0         33300.0   
...               ...                ...             ...   
12175         50055.0            28192.0         18700.0   
12176         70324.0            57025.0         13300.0   
12177         83765.0            55156.0         14800.0   
12178        189930.0            23748.0          7000.0   
12179         48968.0            29786.0          2900.0   

       total_il_high_credit_limit  hardship_flag  debt_settlement_flag  
0                         10000.0              N                     N  
1                         71044.0              N                     N  
2                         46328.0              N                     N  
3                         52017.0              N                     N  
4                         78680.0              N                     N  
...                           ...            ...                   ...  
12175                     19055.0              N

In [5]:
# Data cleansing - finding null values
# print(train_df.isnull().sum())
# print(test_df.isnull().sum())

In [6]:
# Convert categorical data to numeric and separate target feature for training data
X_train_dummies = pd.get_dummies(X_train_df)
X_train_dummies.head()

loan_amnt  int_rate  installment  annual_inc    dti  delinq_2yrs  \
0     7000.0    0.1894       256.38     75000.0  28.62          0.0   
1    40000.0    0.1614       975.71    102000.0  11.72          2.0   
2    11000.0    0.2055       294.81     45000.0  37.25          1.0   
3     4000.0    0.1612       140.87     38000.0  42.89          1.0   
4    14000.0    0.1797       505.93     43000.0  22.16          1.0   

   inq_last_6mths  open_acc  pub_rec  revol_bal  ...  \
0             2.0      20.0      0.0    40414.0  ...   
1             0.0      10.0      0.0    43531.0  ...   
2             3.0      23.0      0.0     8242.0  ...   
3             0.0       7.0      0.0    12767.0  ...   
4             0.0      22.0      0.0    11182.0  ...   

   verification_status_Verified  pymnt_plan_n  initial_list_status_f  \
0                             0             1                      1   
1                             0             1                      0   
2                             1             1                      0   
3                             0             1                      0   
4                             0             1                      0   

   initial_list_status_w  application_type_Individual  \
0                      0                            1   
1                      1                            1   
2                      1                            1   
3                      1                            0   
4                      1                            1   

   application_type_Joint App  hardship_flag_N  hardship_flag_Y  \
0                           0                1                0   
1                           0                1                0   
2                           0                1                0   
3                           1                1                0   
4                           0                1                0   

   debt_settlement_flag_N  debt_settlement_flag_Y  
0                       1                       0  
1                       1                       0  
2                       1                       0  
3                       1                       0  
4                       1                       0  

[5 rows x 92 columns]

In [7]:
# Convert categorical data to numeric and separate target feature for testing data
X_test_dummies = pd.get_dummies(X_test_df)
X_test_dummies.head()

loan_amnt  int_rate  installment  annual_inc    dti  delinq_2yrs  \
0    40000.0    0.1033       856.40    128700.0  12.47          0.0   
1    24450.0    0.1430       572.72     44574.0  15.05          0.0   
2    13500.0    0.1430       316.23     60000.0  28.72          0.0   
3    10625.0    0.1774       268.31     60000.0  15.70          0.0   
4     6375.0    0.1862       232.46     60000.0  35.50          0.0   

   inq_last_6mths  open_acc  pub_rec  revol_bal  ...  \
0             1.0       8.0      0.0    38113.0  ...   
1             1.0       6.0      0.0     1665.0  ...   
2             0.0       8.0      0.0    13857.0  ...   
3             4.0      17.0      0.0     6216.0  ...   
4             0.0      13.0      0.0    12681.0  ...   

   verification_status_Source Verified  verification_status_Verified  \
0                                    1                             0   
1                                    0                             0   
2                                    0                             0   
3                                    0                             1   
4                                    1                             0   

   pymnt_plan_n  initial_list_status_f  initial_list_status_w  \
0             1                      0                      1   
1             1                      0                      1   
2             1                      0                      1   
3             1                      0                      1   
4             1                      0                      1   

   application_type_Individual  application_type_Joint App  hardship_flag_N  \
0                            1                           0                0   
1                            1                           0                1   
2                            1                           0                0   
3                            1                           0                1   
4                            1                           0                1   

   hardship_flag_Y  debt_settlement_flag_N  
0                1                       1  
1                0                       1  
2                1                       1  
3                0                       1  
4                0                       1  

[5 rows x 91 columns]

In [8]:
# To identify missing columns in each of the data set
train_col = list(X_train_dummies.columns)
test_col = list(X_test_dummies.columns)

set1 = set(train_col)
set2 = set(test_col)

missing_set2 = list(sorted(set1 - set2))
missing_set1 = list(sorted(set2 - set1))

print('missing in train_col:', missing_set1)
print('missing in test_col:', missing_set2)

missing in train_col: []
missing in test_col: ['debt_settlement_flag_Y']


In [9]:
# add missing dummy variables to testing set
X_test_dummies['debt_settlement_flag_Y'] = '0'
X_test_dummies

loan_amnt  int_rate  installment  annual_inc    dti  delinq_2yrs  \
0       40000.0    0.1033       856.40    128700.0  12.47          0.0   
1       24450.0    0.1430       572.72     44574.0  15.05          0.0   
2       13500.0    0.1430       316.23     60000.0  28.72          0.0   
3       10625.0    0.1774       268.31     60000.0  15.70          0.0   
4        6375.0    0.1862       232.46     60000.0  35.50          0.0   
...         ...       ...          ...         ...    ...          ...   
4697    30000.0    0.1240       673.42    140480.0  15.74          0.0   
4698    24000.0    0.0756       747.22     50000.0  26.81          0.0   
4699    10000.0    0.2305       387.36     33000.0  38.51          0.0   
4700     8000.0    0.1862       205.86     38000.0  16.36          0.0   
4701    30000.0    0.2055      1123.34    180000.0  12.06          0.0   

      inq_last_6mths  open_acc  pub_rec  revol_bal  ...  \
0                1.0       8.0      0.0    38113.0  ...   
1                1.0       6.0      0.0     1665.0  ...   
2                0.0       8.0      0.0    13857.0  ...   
3                4.0      17.0      0.0     6216.0  ...   
4                0.0      13.0      0.0    12681.0  ...   
...              ...       ...      ...        ...  ...   
4697             0.0      20.0      0.0    23215.0  ...   
4698             0.0       9.0      0.0      459.0  ...   
4699             2.0       7.0      0.0     6342.0  ...   
4700             1.0       8.0      1.0    11636.0  ...   
4701             0.0       8.0      0.0     4771.0  ...   

      verification_status_Verified  pymnt_plan_n  initial_list_status_f  \
0                                0             1                      0   
1                                0             1                      0   
2                                0             1                      0   
3                                1             1                      0   
4                                0             1                      0   
...                            ...           ...                    ...   
4697                             0             1                      1   
4698                             0             1                      0   
4699                             1             1                      1   
4700                             0             1                      0   
4701                             0             1                      1   

      initial_list_status_w  application_type_Individual  \
0                         1                            1   
1                         1                            1   
2                         1                            1   
3                         1                            1   
4                         1                            1   
...                     ...                          ...   
4697                      0                            1   
4698                      1                            1   
4699                      0                            1   
4700                      1                            1   
4701                      0                            1   

      application_type_Joint App  hardship_flag_N  hardship_flag_Y  \
0                              0                0                1   
1                              0                1                0   
2                              0                0                1   
3                              0                1                0   
4                              0                1                0   
...                          ...              ...              ...   
4697                           0                1                0   
4698                           0                1                0   
4699                           0                1                0   
4700                           0                1                0   
4701            

In [10]:
y_train_label = LabelEncoder().fit_transform(train_df['target'])
y_test_label = LabelEncoder().fit_transform(test_df['target'])
# y_test_label

In [11]:
# Train the Logistic Regression model on the unscaled data and print the model score

model = LogisticRegression()

model.fit(X_train_dummies, y_train_label)

print(f'Training Score: {model.score(X_train_dummies, y_train_label)}')
print(f'Testing Score: {model.score(X_test_dummies, y_test_label)}')

Training Score: 0.6530377668308702
Testing Score: 0.5093577201190983


C:\Users\JK\anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:763: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


In [12]:
# Train a Random Forest Classifier model and print the model score

clf = RandomForestClassifier(random_state=1, n_estimators=55).fit(X_train_dummies, y_train_label)
print(f'Training Score: {clf.score(X_train_dummies, y_train_label)}')
print(f'Testing Score: {clf.score(X_test_dummies, y_test_label)}')

Training Score: 1.0
Testing Score: 0.6299447043811144


In [13]:
# Scale the data
scaler = StandardScaler().fit(X_train_dummies)
X_train_scaled = scaler.transform(X_train_dummies)
X_test_scaled = scaler.transform(X_test_dummies)

In [14]:
# Train the Logistic Regression model on the scaled data and print the model score
model = LogisticRegression()

model.fit(X_train_scaled, y_train_label)

print(f'Training Score: {model.score(X_train_scaled, y_train_label)}')
print(f'Testing Score: {model.score(X_test_scaled, y_test_label)}')

Training Score: 0.710919540229885
Testing Score: 0.7598894087622289


C:\Users\JK\anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:763: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


In [15]:
# Train a Random Forest Classifier model on the scaled data and print the model score
clf = RandomForestClassifier(random_state=1, n_estimators=1000).fit(X_train_scaled, y_train_label)
print(f'Training Score: {clf.score(X_train_scaled, y_train_label)}')
print(f'Testing Score: {clf.score(X_test_scaled, y_test_label)}')

Training Score: 1.0
Testing Score: 0.6458953636750319


## Result
As per there above prediction given, Random Forest Classifier performed better then Logistic Regression Model (0.62994 vs 0.50935) as this model produces good prediction with large datasets and without the need to scale the data as the model will try to breakdown the decision tree based on each feature and will have less dependency between the features compared to Logistic Regression Model. This is proven to be true based on the result above where the unscaled data vs scaled data was within the proximity range (0.62994 vs 0.6378136).
On the other hand, logistic regression model have shown significant differences betwee the unscaled vs scaled data where the score is shown 0.5093 vs 0.7598. This is explained due to the unscaled data may provide a biased result due to the magnitude of one feature during the training.